### ISSUE TITLE
 pytorch export to onnx with specific output layers 

### ISSUE THREAD

**Issue by sudapure** [NVIDIA/TensorRT#926]
Am trying to export pytorch network to ONNX and then onnx to tensorrt for faster inferencing , but i need to get the output of layer(avg_pool) prior to final softmax layer or any layer in between, model[-N] , so as per following image, the avgpool layer is named as **GlobalAveragePool**, which returns 512-d feature vector as output tensor, here i am using pretrained resnet-18 model as a sample network backbone.

[![resnet-18 pre-trained model][1]][1]

so my query is how to do this using onnx export script, following is the code which specifies input ,output bindings and exports the model,

```
    model = torchvision.models.resnet18(pretrained=True)
    device = torch.device('cuda:0')
    model.to(device).eval()
    example = torch.rand(1, 3, 128, 128).to(device)
    f = 'saved_weight/weight.onnx'  # filename
    torch.onnx.export(model, example, f, verbose=False, opset_version=12, input_names=['images'],
                              output_names=['output'])
```

But i am unable to figure out how to specify output layer during model export, also following is how i am specifying input,output binding in tensorrt to infer the model and get output tensors,
```
engine, context = build_engine(PLAN_FILE_PATH)
# get sizes of input and output and allocate memory required for input data and for output data
    for binding in engine:
        print(binding)
        if engine.binding_is_input(binding):  # we expect only one input
            input_shape = engine.get_binding_shape(binding)
            # print(input_shape)
            input_size = trt.volume(
                input_shape) * engine.max_batch_size * np.dtype(np.float32).itemsize  # in bytes
            device_input = cuda.mem_alloc(input_size)
        else:  # and one output
            output_shape = engine.get_binding_shape(binding)
            # print(output_shape)
            # create page-locked memory buffers (i.e. won't be swapped to disk)
            host_output = cuda.pagelocked_empty(trt.volume(
                output_shape) * engine.max_batch_size, dtype=np.float32)
            device_output = cuda.mem_alloc(host_output.nbytes)

# preprocess input data
host_input = np.array(preprocess_image(img_path),dtype=np.float32, order='C')
print(host_input.shape)
cuda.memcpy_htod_async(device_input, host_input, stream)

# run inference
context.execute_async(bindings=[int(device_input), int(device_output)],stream_handle=stream.handle)

```
*So at the time of infer using tensorrt the output bindings would be the one which had been specified at the time of exporting the model, so in case of having multiple output we can simply specify output binding to get the tensors for specified layer.*

  [1]: https://i.stack.imgur.com/JpcNv.png

**Comment by ttyio** [comment]
Hello @sudapure , are you looking for method to mark the `globalaveragePool` as output?
We have 2 different approach here:
One is to modify the `resnet.py` in torchvision source, change the `forward` function to also export this tensor as function return;
The other is to use onnxGraphSurgeon to mark this `globalaveragePool` as output, you can find the tool in https://github.com/NVIDIA/TensorRT/tree/master/tools/onnx-graphsurgeon

**Comment by ttyio** [comment]
Closing since no activity in this thread for more than 3 weeks, please reopen if you still have question, thanks!
